core/org.wso2.carbon.tomcat.ext/src/main/java/org/wso2/carbon/tomcat/ext/valves/RequestCorrelationIdValve.java

```diff
@@ -0,0 +1,338 @@
+import org.apache.log4j.MDC;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.http.HttpServletRequest;
+
+/**
+ * Tomcat valve which adds MDC from a header value received.
+ * The header and its MDC can be configured.
+ * <p>
+ * By default a HTTP header "activityid" is added to MDC "Correlation-ID"
+ * <p>
+ * The header and MDC can be configured in tomcat valve configuration like,
+ * <code>
+ * headerToCorrelationIdMapping={'activityid':'Correlation-ID'}
+ * queryToCorrelationIdMapping={'RelayState':'Correlation-ID'}
+ * </code>
+ */
+public class RequestCorrelationIdValve extends ValveBase {
+
+    private static final Log correlationLog = LogFactory.getLog("correlation");
+    private static final String CORRELATION_ID_MDC = "Correlation-ID";
+    private Map<String, String> headerToIdMapping;
+    private Map<String, String> queryToIdMapping;
+    private static List<String> toRemoveFromThread = new ArrayList<>();
+    private String correlationIdMdc = CORRELATION_ID_MDC;
+    private String headerToCorrelationIdMapping;
+    private String queryToCorrelationIdMapping;
+    private String configuredCorrelationIdMdc;
+    private static final String CORRELATION_LOG_REQUEST_START = "HTTP-In-Request";
+    private static final String CORRELATION_LOG_SEPARATOR = "|";
+    private static final String CORRELATION_LOG_REQUEST_END = "HTTP-In-Response";
+    private static final String CORRELATION_LOG_SYSTEM_PROPERTY = "enableCorrelationLogs";
+
+    @Override
+    protected void initInternal() throws LifecycleException {
+
+        super.initInternal();
+        Gson gson = new Gson();
+        if (StringUtils.isNotEmpty(headerToCorrelationIdMapping)) {
+            headerToIdMapping = gson.fromJson(this.headerToCorrelationIdMapping, Map.class);
+            toRemoveFromThread.addAll(headerToIdMapping.values());
+        }
+
+        if (StringUtils.isNotEmpty(queryToCorrelationIdMapping)) {
+            queryToIdMapping = gson.fromJson(this.queryToCorrelationIdMapping, Map.class);
+            toRemoveFromThread.addAll(queryToIdMapping.values());
+        }
+
+        if (StringUtils.isNotEmpty(configuredCorrelationIdMdc)) {
+            correlationIdMdc = configuredCorrelationIdMdc;
+        }
+    }
+
+    @Override
+    public void invoke(Request request, Response response) throws IOException, ServletException {
+
+        long requestStartTime = System.currentTimeMillis();
+        try {
+
+            Map<String, String> associateToThreadMap = new HashMap<>(getHeadersToAssociate(request));
+            if (associateToThreadMap.size() == 0) {
+                associateToThreadMap.putAll(getQueryParamsToAssociate(request));
+            }
+
+            if (associateToThreadMap.size() == 0) {
```

Comment:
if conditions in line 96 and 101  are same. Can group together

Reply:
the logic here is first we try to associate from the headers, if there are is nothing in the headers then we try to associate from the query params, if it is also not set then we generate a new random UUID. so reason being for two if conditions is that only one of these would happen.

as mentioned in https://github.com/wso2-support/carbon4-kernel/pull/505#discussion_r303734717